File: memallocspeed/memallocspeed.go
// for https://groups.google.com/forum/#!topic/golang-nuts/HZ_sBoLtnX4
package main

import (
	"fmt"
	"time"
)

func domemalloc() {
	defer func() {
		if err := recover(); err != nil {
			println("panic for", err)
		}
	}()
	slimslice := make([]uint64, 256, 256)
	fatslice := make([]uint64, 0, 1024*1024*1024/8)
	start := time.Now()
	for {
		fatslice = append(fatslice, slimslice...)
		if len(fatslice)+len(slimslice) > cap(fatslice) {
			size := len(fatslice) * 8
			t := time.Since(start).Seconds()
			fmt.Printf("%d MB at %.2f = %.2f MB/s\n", size>>20, t, float64(size)/(1<<20)/t)
		}
	}
}
func main() {
	domemalloc()
}

File: memallocspeed/memallocspeed.c
// for https://groups.google.com/forum/#!topic/golang-nuts/HZ_sBoLtnX4 
#include <stdio.h>
#include <sys/time.h>
#include <stdlib.h>
#include <string.h>

char zero[256*8];

double
now(void)
{
    struct timeval tv;
    gettimeofday(&tv, 0);
    return (double)tv.tv_sec+tv.tv_usec/1e6;
}

int 
main(void)
{
    char *p;
    long long size, i;
    double start, t;

    start = now();
    size = 1<<30;
    for(;;) {
        p = malloc(size);
        if (p==NULL) {
                printf("malloc failed, press <CTL+C> to exit.\n");
                getchar();
                return 1;
        }
        for(i=0; i<size; i+=sizeof zero){
            memmove(p+i, zero, sizeof zero);
        }
        t = now() - start;
        printf("%lld MB at %.2f = %.2f MB/s\n", size>>20, t, size/(double)(1<<20)/t);
        size += size/4;
    }

    return 0;
}
